4.next: Allow Route Path params inside string representation

LordSimal opened the pull request:
Currently our `parseRoutePath()` method inside the Router only allows Routes to be represented like
```
[Plugin.][Prefix/]Controller::action
```

But it may easily be the case, that Routes need to be represented with action params as well like
```
[Plugin.][Prefix/]Controller::action/param1/param2/param3
```

This PR adds exactly this functionality.

I also fixed the Regex to properly escape `/` characters which is recommended even if they are inside a capture group.

Are there any additional edge-cases I am not thinking of? Happy to get some feedback 😁

LordSimal committed:
add route path params feature

ADmad commented on src/Routing/Router.php:
```diff
@@ -1010,11 +1010,12 @@ public static function parseRoutePath(string $url): array
             (?<controller>[a-z0-9]+)
             ::
             (?<action>[a-z0-9_]+)
+            (?<params>[a-z0-9_\/]*)
```
I don't think escaping the `/` is necessary since the regex uses `#` as delimiters not `/`.

LordSimal replied:
Ah.... so thats what this does at the start... Wasn't aware that one can set different regex delimiters 😅 
Will revert that back to how it was.

Ggizzle replied:
Great idea 💡😫

Ggizzle replied:
> Ah.... so thats what this does at the start... Wasn't aware that one can set different regex delimiters 😅 
> 
> Will revert that back to how it was.

What I think is 🤔.. Is gonna always gonna be problematic.. I have a Good ideas 💡 for it

ADmad commented on tests/TestCase/Routing/RouterTest.php:
```diff
@@ -2260,6 +2260,24 @@ public function testParseRoutePath(): void
+        $this->assertSame($expected, Router::parseRoutePath('Bookmarks::view/1'));
+        $this->assertSame($expected, Router::parseRoutePath('Bookmarks::view/1/'));
+
+        $expected = [
```
Such an array would only be useful for routes ending with `/*` or `/**`, since for named route elements the element name is required as key. I am not sure we should be providing features which only work for a subset of routes.

Another alternative would be for e.g.  supporting string like `Cms.Articles::edit/year:2023/month:03/day:5th` which gets parsed into array like:
```
[
    'plugin' => 'Cms',
    'controller' => 'Articles',
    'action' => 'edit',
    'year' => '2023',
    'month' => '03',
    'day' => '5th',
]
```

LordSimal replied:
how would this named arguments approach work with something like the PagesController::display method?
I e.g. have a Page URL which looks like `pages/statistic/2023/03` and therefore would require something like 
```
Pages::display/statistic/2023/03
```

ADmad replied:
Routes that don't have named element won't use any :)

ADmad replied:
I haven't gone through the code (or actually used it) but Croogo has an implementation of something similar:
https://github.com/croogo/croogo/blob/fc0648659dcb5790ba8a7429250dae492ca724ec/Core/src/Utility/StringConverter.php#L97

LordSimal replied:
I have adapted the regex to support "named arguments" kind of strings even though I am not sure which other string parts can be parts of the "path"

markstory replied:
> Another alternative would be for e.g. supporting string like Cms.Articles::edit/year:2023/month:03/day:5th which gets parsed into array like:

Would we have any rules on what order parameters of different kinds are defined? For example can you put numeric indexed parameters before *and* after named ones? eg `Articles::view/param/key:value/param2`?

Should we be concerned that the `key:value` syntax looks the same as 'named parameters' which is a feature we removed in 3.x? Should we use `key=value` so that the syntax is closer to query string parameters? Or is that more confusing?

LordSimal replied:
I am fine with either approach since we are manually processing the "path" part of this "URL" anyways.

LordSimal replied:
I'll try the `key=value` approach later today

ADmad replied:
> Should we be concerned that the key:value syntax looks the same as 'named parameters' which is a feature we removed in 3.x?

That's a valid concern. Reusing the separator char we deprecated could be confusing for users.

LordSimal replied:
The only part I am not sure is which kind of characters we should allow to be part of the `key=value` part.
Currently we restrict it to `[a-z0-9_/:=]`

ADmad replied:
The key regex should be same as https://github.com/cakephp/cakephp/blob/cf67e6148943d9d67025e697ec240a0a91b2720e/src/Routing/Route/Route.php#L116 sans the braces.

LordSimal replied:
Wasn't able to just use it directly but I hopeyfully adjusted it now to what it should be ^^

LordSimal committed:
add named arguments support

dereuromark commented on tests/TestCase/Routing/RouterTest.php:
```diff
@@ -2260,6 +2260,40 @@ public function testParseRoutePath(): void
+        $expected = [
+            'controller' => 'Bookmarks',
+            'action' => 'view',
+            'organisation' => 'cakephp',
```
Nitpick: should be z since we use AE

LordSimal replied:
my german speaking brain is just getting in my ways sometimes 😁

LordSimal committed:
fix typo

LordSimal committed:
change key:value to key=value

LordSimal committed:
adjust regex

LordSimal commented:
Anything missing here or can this be merged?

ADmad commented on src/Routing/Router.php:
```diff
@@ -1032,6 +1033,20 @@ public static function parseRoutePath(string $url): array
         $defaults['action'] = $matches['action'];
 
+        if ($matches['params'] !== '') {
+            $paramsArray = array_values(array_filter(explode('/', $matches['params'])));
```
Is the `array_values()` necessary? Could the `array_filter()` be avoided if we `rtim()` the trailing slash instead?

LordSimal replied:
indeed

LordSimal committed:
replace array_* methods with trim

garas commented on src/Routing/Router.php:
```diff
@@ -1015,6 +1015,7 @@ public static function parseRoutePath(string $url): array
             (?<controller>[a-z0-9]+)
             ::
             (?<action>[a-z0-9_]+)
+            (?<params>[a-z0-9-_/=]*)
```
```suggestion
            (?<params>(?:/(?:[a-z][a-z0-9-_]*=)?[a-z0-9-_]+)+/?)
```
These won't be accepted anymore:
```
Bookmarks::view-1
Bookmarks::view=1
Bookmarks::view/=value
Bookmarks::view/key=
Bookmarks::view/key=value=value
Bookmarks::view/0key=value
Bookmarks::view/-key=value
```

LordSimal replied:
@markstory you want to have tests added for those kind of scenarios?

LordSimal replied:
actually this breaks other tests as well so I will revert that commit

garas replied:
```suggestion
            (?<params>(?:/(?:[a-z][a-z0-9-_]*=)?[a-z0-9-_]+)+/?)?
```
I've forgot `?` in end to make params optional. I've tested only with params, so I missed cases without params.

markstory replied:
> you want to have tests added for those kind of scenarios?

I don't think we need to have tests for all the invalid data values. But we should have coverage on the valid states that we expect users to use. I think we should not allow keys with leading numbers as they can't be turned into PHP variables with passed parameters. While they are valid outside of that context, we shouldn't have incompatible features within the framework.

LordSimal replied:
Found (another) regex to validate the key string is a valid variable name.

LordSimal replied:
The only thing I am struggeling with right now is the `Could not parse a key=value from` exception. Can't find a good example where this exception would throw so I am thinking of just removing that check...

LordSimal committed:
Update regex

Co-authored-by: Edgaras Janušauskas <edgaras.janusauskas@gmail.com>

LordSimal committed:
Revert "Update regex"

This reverts commit 2f9501a0a50fa8441b805adc019f996d66b4d9dd.

LordSimal committed:
Use more restrictive regex

Co-authored-by: Edgaras Janušauskas <edgaras.janusauskas@gmail.com>

LordSimal committed:
fix tests

Ggizzle commented:
What I. Think 🤔 Is Unplug and Reset 🧐

Ggizzle commented:
🤔 reset modomn

markstory commented on tests/TestCase/Routing/RouterTest.php:
```diff
@@ -2260,6 +2260,40 @@ public function testParseRoutePath(): void
+            'organization' => 'cakephp',
+            'bake',
+        ];
+        $this->assertSame($expected, Router::parseRoutePath('Bookmarks::view/organization=cakephp/bake'));
```
How do we support `=` **in** a parameter? I suspect that userland code will want to format values into route strings and that will eventually include `=`. Could we have an escaped or quoted value?

markstory commented on src/Routing/Router.php:
```diff
@@ -1015,6 +1015,7 @@ public static function parseRoutePath(string $url): array
             (?<controller>[a-z0-9]+)
             ::
             (?<action>[a-z0-9_]+)
+            (?<params>(?:/(?:[a-z][a-z0-9-_]*=)?[a-z0-9-_]+)+/?)?
```
```suggestion
            (?<params>(?:\/(?:[a-z][a-z0-9-_]*=)?(["']?[a-z0-9-_]+['"]?))+\/?)?
```

I think this makes it possible to parse the quotes out. It will require slightly more complex parsing logic. I'm happy to take care of that if you want.

LordSimal replied:
I see what you mean but if we want to support something like
```
Bookmarks::view/organization="cakephp=test"
```
then it should be this regex in my opinion (since we already use a custom regex delimiter and need to escape the `'`) 
```
(?<params>(?:/(?:[a-z][a-z0-9-_]*=)?(["\']?[a-z0-9-_=]+[\'"]?))+/?)?
```
But as you already said the splitting of `key=value` is more complex... Need to look into it.

LordSimal committed:
add regex for param parsing

LordSimal committed:
prevent invalid param key names

garas commented on src/Routing/Router.php:
```diff
@@ -1042,7 +1042,12 @@ public static function parseRoutePath(string $url): array
                     }
-                    $convertedArray[$paramMatches['key']] = trim(trim($paramMatches['value'], '"'), "'");
+                    $paramKey = $paramMatches['key'];
+                    if (!preg_match('/^[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*$/', $paramKey)) {
```
Key is already checked and limited to `[a-z][a-z0-9-_]*`.

It same as placeholder regex 
https://github.com/cakephp/cakephp/blob/550fbc51919ab5cc6ff135814fee8cca831e52d0/src/Routing/Route/Route.php#L116

LordSimal replied:
The `params` match from the inital regex currently does match "invalid keys" aka invalid `key=value` pairs where the `key` can't be extracted to a PHP variable (thats how I understood Mark Story at least).
So `Bookmarks::view/-invalid=cakephp/1otherinvalid=cakephp` would result in a `params` of 
```
/-invalid=cakephp/1otherinvalid=cakephp
```
Which then gets pased down and needs to be re-evaluated.

markstory replied:
> pairs where the key can't be extracted to a PHP variable (thats how I understood Mark Story at least).

Yeah that was my concern. Another angle on this problem could be that we don't worry so much about the keys as users will get errors from the framework elsewhere indicating their mistake. However, I think quoted values are something we need as `=` is valid within a path segment

LordSimal replied:
I also agree that having `value` data which itself contains `=` is a likely scenario.

markstory committed:
Improve parseRoutePath tests

- Use a dataprovider for successful cases.
- Separate failure cases into separate test methods as you can only have
  one expectException() per test method.

markstory committed:
Improve efficiency of route parsing and quoted values

- Remove additional allocations, operations, and temporary arrays.
- only need to trim once.
- limit memory consumption for route path now that users can put
  arbitrary data into it.

markstory commented:
I pushed up a couple changes to:

- Make tests use a data provider. I find this makes expanding tests in the future easier.
- Fix the failure tests as only the first one was running before.
- Expand quoted values to accept symbols and anything that isn't a quote.
- Constrained the memory footprint of the URL parse cache as users can start sending arbitrary data to this function which would mean unbounded memory usage.

I think this looks good to go now.

LordSimal commented:
I'd say if there are no objections from anyone else we can merge this.
